homeassistant/components/alarm_control_panel/mqtt.py

```diff
@@ -46,9 +49,26 @@
 }).extend(mqtt.MQTT_AVAILABILITY_SCHEMA.schema)
 
 
-@asyncio.coroutine
-def async_setup_platform(hass, config, async_add_entities,
-                         discovery_info=None):
+async def async_setup_platform(hass: HomeAssistantType, config: ConfigType,
+                               async_add_entities, discovery_info=None):
+    """Set up MQTT alarm control panel through configuration.yaml."""
+    await _async_setup_platform(hass, config, async_add_entities,
+                                discovery_info)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up MQTT alarm control panel dynamically through MQTT discovery."""
+    async def async_discover(config):
+        """Discover and add an MQTT alarm control panel."""
+        await _async_setup_platform(hass, {}, async_add_entities, config)
+
+    async_dispatcher_connect(
+        hass, MQTT_DISCOVERY_NEW.format(alarm_control_panel.DOMAIN, 'mqtt'),
+        async_discover)
+
+
+async def _async_setup_platform(hass, config, async_add_entities,
```

Comment:
It should just be called `_async_setup_entity(hass, async_add_entities, config)` (in all the platforms).

Inside `async_setup_platform` you know that it's always config and no longer discovery_info (as that's moved to config entries). In the setup entry discover, you know that `config` is always the config.